### PR TITLE
[runtime] Deny loading bad images.

### PR DESCRIPTION
**Pull request by kumpera**
This allows us to prioritize system libs over current-dir libs.

### CODE REVIEW

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -1898,6 +1898,30 @@ real_load (gchar **search_path, const gchar *culture, const gchar *name, gboolea
 	return result;
 }
 
+//Keep it ordered
+static const char *priority_assemblies[] = {
+	"System.Runtime.InteropServices.RuntimeInformation",
+};
```
Comment (marek-safar) [inline mono/metadata/appdomain.c]: Here is a list of additional required assemblies

System.Globalization.Extensions.dll
System.IO.Compression.dll
System.Net.Http.dll
System.Text.Encoding.CodePages.dll

**Comment by marek-safar** [comment]
As you can see from the list of assemblies we need to **somehow** identify that the assembly is the one we want to replace and it's not the one user explicitly put there. For assemblies like **System.Net.Http** this can be quite common that user has local copy which has a fix not yet released by mono and wants that assembly to be used. Similarly when we build mono I want the newly build assembly to be used instead of GAC one.

As this is such gross hack breaking .net rules we need to add logging that it's happening.

Another thing we should do is to check that the GAC assembly is actually there otherwise we could break another scenario when someone runs mono without GAC

**Comment by kumpera** [comment]
Hey Marek,

On your comments.

How would I identify that an assembly is one we'd like replaced?

This change doesn't affect GAC lookup, it simply swaps the lookup order between current dir and profile dir for specific assembly names.

Logging is already in place for assembly loading. It tells where it was loaded from.

On this possibly breaking the GAC, I don't understand how this would be affected.

**Comment by kumpera** [comment]
We had a discussion and agreed that this will need a different approach to work.

We need to check the matched assembly by name + mvid before considering whether to not load it.

**Comment by kumpera** [comment]
I rewrote the approach to deny loading the bad assemblies.
This is a much more surgical change as it matches on file name and GUID.

Meaning that we'll need to add multiple entries for the popular nugets with multiple versions, like System.Net.Http.

File: mono/metadata/image.c [hunk 2]
```diff
@@ -1077,6 +1077,69 @@ install_pe_loader (void)
...
+        h = ((h << 5) + h) ^ *str;
+		++str;
+	}
+
+	return h;
+}
+
+static gboolean
+is_problematic_image (MonoImage *image)
+{
```
Comment (vargaz) [inline mono/metadata/image.c]: Wouldn't be easier to just put them in a GHashTable ?

Comment (kumpera) [inline mono/metadata/image.c]: We'd have to build that hashtable on startup, not a big fan of that either.

File: tools/nuget-dump/nuget-dump.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	static void DumpNuget (string nupkg) {
+		var zip = new ZipArchive(new FileStream (nupkg, FileMode.Open));
+
+		var nuspec = FindSpecFile (zip);
+		var l = XElement.Load (new StreamReader (nuspec.Open ()));
+		var version = (from el in l.Descendants() where el.Name.LocalName == "version" select el.Value).FirstOrDefault ();
+
+		foreach (var et in from e in zip.Entries where e.FullName.StartsWith ("lib") && e.Name.EndsWith (".dll") select e) {
```
Comment (marek-safar) [inline tools/nuget-dump/nuget-dump.cs]: We are interested in dlls from folders `net*` only

File: tools/nuget-dump/nuget-dump.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
+using System;
```
Comment (marek-safar) [inline tools/nuget-dump/nuget-dump.cs]: Please choose better name e.g. nuget2ignored or nuget-hash-extractor or something like that

File: mono/metadata/image.c [hunk 5]
```diff
@@ -1087,16 +1091,25 @@ typedef struct {
...
 
-static const ProblematicAssembly bad_assemblies [1] = {
-	{
-		.hash = 0xD07383BB,
-		.assembly_name = SYS_RT_INTEROP_RUNTIME_INFO,
-		"DD91439F-3167-478E-BD2C-BF9C036A1395"
-	}
+#define BAD_ASSEMBLY(HASH, NAME, GUID, VER_STR)	{ .hash = HASH, .assembly_name = NAME, .guid = GUID }
+
+static const ProblematicAssembly bad_assemblies [] = {
```
Comment (marek-safar) [inline mono/metadata/image.c]: Bad sounds bad ;-)
What about ignored_assemblies instead ?

Comment (marek-safar) [inline mono/metadata/image.c]: We need more versions that just the latest one e.g. for System.Net I think we need at least 4.3.0, 4.1.1, 4.1.0, 4.0.0

note: 4.0.0.0 is the most popular

Comment (akoeplinger) [inline mono/metadata/image.c]: We should also include version numbers for the upcoming .NET Core 2.0 release, most/all of them should be 4.4.0

Comment (marek-safar) [inline mono/metadata/image.c]: I don't think we have final MVID so that will have to wait

Comment (akoeplinger) [inline mono/metadata/image.c]: Hmm, right 😞

File: tools/nuget-dump/nuget-dump.cs (hunk 4 above)
Comment (kumpera) [inline tools/nuget-dump/nuget-dump.cs]: done

**Comment by kumpera** [comment]
Failures look unrelated.

**Comment by ludovic-henry** [comment]
This breaks windows build: https://jenkins.mono-project.com/job/w/4049/parsed_console/log_content.html#ERROR1 and https://jenkins.mono-project.com/job/x/2237/parsed_console/log_content.html#ERROR1

File: mono/metadata/image.c [hunk 6]
```diff
@@ -1077,6 +1077,102 @@ install_pe_loader (void)
...
+	"System.Globalization.Extensions.dll",
+	"System.IO.Compression.dll",
+	"System.Net.Http.dll",
+	"System.Text.Encoding.CodePages.dll"
+};
+
+#define IGNORED_ASSEMBLY(HASH, NAME, GUID, VER_STR)	{ .hash = HASH, .assembly_name = NAME, .guid = GUID }
+
+static const IgnoredAssembly ignored_assemblies [] = {
+	IGNORED_ASSEMBLY (0x1136045D, SYS_GLOBALIZATION_EXT, "475DBF02-9F68-44F1-8FB5-C9F69F1BD2B1", "4.0.0 net46"),
```
Comment (jaredpar) [inline mono/metadata/image.c]: The GUID portion represents the MVID of the assembly correct?  

I don't believe that CoreFX uses deterministic builds yet hence this GUID would be different if the same sources were built manually.  That would seem to compromise the ability of Mono to handle this scenario on these builds.  

I assume that's okay but wanted to know if it was understood.

Comment (akoeplinger) [inline mono/metadata/image.c]: @jaredpar yes, it's the MVID. This is a very targeted hack and we do understand it doesn't catch all cases (e.g. preview packages from myget etc).

We're still waiting for a proper solution in corefx: https://github.com/dotnet/corefx/issues/15112

**Pull request merged**